# Pandas for Data Science
The pandas module contains data structures and data manipulation tools designed to make data cleaning and analysis fast and easy in Python.

<br>
<br>
<br>

## Basics:

### Install Pandas:
To install pandas in your system, run the below command.

### Import pandas
Import pandas with an alias 'pd'

In [1]:
import pandas as pd

<br>
<br>
<br>

## Pandas Data Structures
pandas provides 2 important data structures:
1. **Series**: a one-dimensional array-like object containing a sequence of values and an associated array of data labels, called its index.
2. **DataFrame**: a rectangular table of data and contains an ordered collection of columns, each of which can be a different value type

### Creating pandas series:

In [2]:
# using the pd.Series() function:
data= pd.Series([1,2,3,4,5,6])
data

0    1
1    2
2    3
3    4
4    5
5    6
dtype: int64

The pd.Series() function accepts any type of i dimentional sequence object and creates a pandas Series with its elements.

In [3]:
# specifying the index
data2= pd.Series(['Alice','Bob','Claire','Dean','Elaina'],index=['a','b','c','d','e'])
data2

a     Alice
b       Bob
c    Claire
d      Dean
e    Elaina
dtype: object

The pd.Series() function also provides a way to specify the indices of the series object. If not specified, the elements will be indexed starting from 0 by default.

### Creating pandas DataFrame:

#### Using Lists

In [4]:
df=pd.DataFrame([['Name','Age','Grade'],['Alice',20,'A'],['Bob',19,'B'],['Claire',19,'A'],['Dean',20,'B']])
print(df)

        0    1      2
0    Name  Age  Grade
1   Alice   20      A
2     Bob   19      B
3  Claire   19      A
4    Dean   20      B


A DataFrame can be created using Lists of lists, where, each list each considered as an individual row. 

In [7]:
# Specifying column names
df=pd.DataFrame([['Alice',20,'A'],['Bob',19,'B'],['Claire',19,'A'],['Dean',20,'B']],columns=['Name','Age','Grade'])
print(df)

     Name  Age Grade
0   Alice   20     A
1     Bob   19     B
2  Claire   19     A
3    Dean   20     B


The pd.DataFrame() function allows specification of column and row names of the object

#### Using Dictionaries

In [8]:
df=pd.DataFrame({'Name':['Alice','Bob','Claire','Dean'],'Age':[19,20,19,19],'Grade':['A','B','A','B']})
print(df)

     Name  Age Grade
0   Alice   19     A
1     Bob   20     B
2  Claire   19     A
3    Dean   19     B


While creating a DataFrame using Dictionaries, each item becomes an individual column.

#### Using CSV files

In [57]:
data=pd.read_csv("Python Programming Lab.csv")
df=pd.DataFrame(data)
df

S No  Lab 1  Lab 2  Lab 3  Lab 4  Lab 5  Lab 6  Lab 7  Lab 8\n  Lab 9  \
0      1    NaN   80.0  100.0    NaN  100.0  100.0  100.0      NaN    NaN   
1      2  100.0  100.0  100.0  100.0    NaN  100.0    NaN     50.0  100.0   
2      3  100.0   95.0  100.0  100.0  100.0  100.0  100.0    100.0  100.0   
3      4   80.0  100.0  100.0  100.0  100.0  100.0  100.0    100.0   50.0   
4      5   80.0  100.0  100.0  100.0    NaN  100.0  100.0    100.0   50.0   
..   ...    ...    ...    ...    ...    ...    ...    ...      ...    ...   
65    66    NaN    NaN  100.0  100.0   90.0  100.0  100.0      NaN    NaN   
66    67    NaN  100.0    NaN  100.0  100.0  100.0  100.0      NaN    NaN   
67    68    NaN  100.0  100.0  100.0  100.0  100.0  100.0     80.0   50.0   
68    69    NaN    NaN    NaN  100.0  100.0  100.0  100.0      NaN   80.0   
69    70    NaN  100.0  100.0  100.0  100.0  100.0    NaN      NaN    NaN   

    Lab 10  Lab 11  Lab Total  Lab Marks  Record  Lab Internal  Attendance  \
0      NaN     NaN        480          7     8.0            13           4   
1     90.0    90.0        830         12     8.0            18           4   
2    100.0     NaN        995         14     8.0            19           5   
3    100.0     NaN        930         13     7.0            16           5   
4      NaN     NaN        730         10     8.0            14           4   
..     ...     ...        ...        ...     ...           ...         ...   
65    80.0    85.0        655          9     7.0            13           2   
66     NaN     NaN        500          7     7.0            16           4   
67    80.0     NaN        810         12     7.0            12           4   
68    10.0     NaN        490          7     7.0            12           2   
69     NaN     NaN        500          7     NaN            17           2   

    Total  
0      32  
1      42  
2      46  
3      41  
4      36  
..    ...  
65     31  
66     34  
67     35  
68     28  
69     26  

[70 rows x 18 columns]

<br>
<br>
<br>

## Data Handling using pandas

Let us use the above csv file to understand how pandas is helpful in Data Science. The data contains marks of 70 students in various areas.
The main context in the data:
- Each Student has performed 11 lab sessions. The total marks scored in all of these sessions is reduced to 15.
- Each student also has marks alloted for Records, Internals and attendance.
- The total marks include the sum of all these data.


### Acessing Data in a DataFrame 

The data can be acessed using basic indexing or by using the column names. To acess data of individual rows, pandas provied loc() and iloc() functions.

In [58]:
# Total marks of all students
df.Total

0     32
1     42
2     46
3     41
4     36
      ..
65    31
66    34
67    35
68    28
69    26
Name: Total, Length: 70, dtype: int64

In [59]:
df['Total']

0     32
1     42
2     46
3     41
4     36
      ..
65    31
66    34
67    35
68    28
69    26
Name: Total, Length: 70, dtype: int64

In [60]:
# lab marks and total marks
df[['Lab Marks','Total']]

Lab Marks  Total
0           7     32
1          12     42
2          14     46
3          13     41
4          10     36
..        ...    ...
65          9     31
66          7     34
67         12     35
68          7     28
69          7     26

[70 rows x 2 columns]

#### Using the loc and iloc functions

The loc function is used to acess individual rows based on the row names. The iloc function is used for acessing with row indices.

In [61]:
df.loc[34]

S No              35.0
Lab 1            100.0
Lab 2            100.0
Lab 3            100.0
Lab 4            100.0
Lab 5            100.0
Lab 6            100.0
Lab 7            100.0
Lab 8\n           50.0
Lab 9             80.0
Lab 10            95.0
Lab 11            90.0
Lab Total       1015.0
Lab Marks         14.0
Record             8.0
Lab Internal      18.0
Attendance         5.0
Total             45.0
Name: 34, dtype: float64

In [62]:
df.loc[[34,35,36]]

S No  Lab 1  Lab 2  Lab 3  Lab 4  Lab 5  Lab 6  Lab 7  Lab 8\n  Lab 9  \
34    35  100.0  100.0  100.0  100.0  100.0  100.0  100.0     50.0   80.0   
35    36  100.0  100.0  100.0  100.0  100.0  100.0  100.0     80.0   90.0   
36    37  100.0  100.0  100.0  100.0  100.0  100.0  100.0    100.0    NaN   

    Lab 10  Lab 11  Lab Total  Lab Marks  Record  Lab Internal  Attendance  \
34    95.0    90.0       1015         14     8.0            18           5   
35    80.0    90.0       1040         15     9.0            18           5   
36    80.0    90.0        970         14     7.0            14           5   

    Total  
34     45  
35     47  
36     40

#### Conditional Accessing:

We can get only the specific rows that satisfy a codition using the loc function. For example, let us try to filter the data to display only those records whose total is greater than 45.

In [63]:
df.loc[df['Total']>45]

S No  Lab 1  Lab 2  Lab 3  Lab 4  Lab 5  Lab 6  Lab 7  Lab 8\n  Lab 9  \
2      3  100.0   95.0  100.0  100.0  100.0  100.0  100.0    100.0  100.0   
9     10   90.0  100.0  100.0  100.0  100.0  100.0  100.0    100.0  100.0   
28    29  100.0  100.0  100.0  100.0  100.0  100.0  100.0     80.0   50.0   
35    36  100.0  100.0  100.0  100.0  100.0  100.0  100.0     80.0   90.0   
58    59   90.0  100.0  100.0  100.0  100.0  100.0  100.0    100.0   90.0   

    Lab 10  Lab 11  Lab Total  Lab Marks  Record  Lab Internal  Attendance  \
2    100.0     NaN        995         14     8.0            19           5   
9     95.0    80.0       1065         15     9.0            19           5   
28    70.0    80.0        980         14     8.0            19           5   
35    80.0    90.0       1040         15     9.0            18           5   
58    90.0    85.0       1055         15     9.0            19           5   

    Total  
2      46  
9      48  
28     46  
35     47  
58     48

### Handling Missing Data:
pandas provides the isnull(), dropna() and fillna() functions to handle missing data. 

The isnull() function provies a similar shaped matrix that contains true if the cell value is null and false if not. The notnull() function is the opposite of isnull().

In [64]:
df.isnull()

S No  Lab 1  Lab 2  Lab 3  Lab 4  Lab 5  Lab 6  Lab 7  Lab 8\n  Lab 9  \
0   False   True  False  False   True  False  False  False     True   True   
1   False  False  False  False  False   True  False   True    False  False   
2   False  False  False  False  False  False  False  False    False  False   
3   False  False  False  False  False  False  False  False    False  False   
4   False  False  False  False  False   True  False  False    False  False   
..    ...    ...    ...    ...    ...    ...    ...    ...      ...    ...   
65  False   True   True  False  False  False  False  False     True   True   
66  False   True  False   True  False  False  False  False     True   True   
67  False   True  False  False  False  False  False  False    False  False   
68  False   True   True   True  False  False  False  False     True  False   
69  False   True  False  False  False  False  False   True     True   True   

    Lab 10  Lab 11  Lab Total  Lab Marks  Record  Lab Internal  Attendance  \
0     True    True      False      False   False         False       False   
1    False   False      False      False   False         False       False   
2    False    True      False      False   False         False       False   
3    False    True      False      False   False         False       False   
4     True    True      False      False   False         False       False   
..     ...     ...        ...        ...     ...           ...         ...   
65   False   False      False      False   False         False       False   
66    True    True      False      False   False         False       False   
67   False    True      False      False   False         False       False   
68   False    True      False      False   False         False       False   
69    True    True      False      False    True         False       False   

    Total  
0   False  
1   False  
2   False  
3   False  
4   False  
..    ...  
65  False  
66  False  
67  False  
68  False  
69  False  

[70 rows x 18 columns]

To get the number of null values in each column:

In [65]:
df.isnull().sum()

S No             0
Lab 1           17
Lab 2            3
Lab 3            7
Lab 4            3
Lab 5            8
Lab 6            4
Lab 7            8
Lab 8\n         15
Lab 9           21
Lab 10          15
Lab 11          36
Lab Total        0
Lab Marks        0
Record           3
Lab Internal     0
Attendance       0
Total            0
dtype: int64

To get total null values in the data frame:

In [66]:
df.isnull().sum().sum()

140

For efficient analysis, handling missing data is crucial. Let try to eliminate missing values by filling them with the means of the respective columns.

In [67]:
df['Lab 1']=df['Lab 1'].fillna(100)

In [68]:
df=df.fillna(df.mean())
df

S No  Lab 1       Lab 2      Lab 3       Lab 4       Lab 5  Lab 6  \
0      1  100.0   80.000000  100.00000   99.552239  100.000000  100.0   
1      2  100.0  100.000000  100.00000  100.000000   99.193548  100.0   
2      3  100.0   95.000000  100.00000  100.000000  100.000000  100.0   
3      4   80.0  100.000000  100.00000  100.000000  100.000000  100.0   
4      5   80.0  100.000000  100.00000  100.000000   99.193548  100.0   
..   ...    ...         ...        ...         ...         ...    ...   
65    66  100.0   99.328358  100.00000  100.000000   90.000000  100.0   
66    67  100.0  100.000000   99.84127  100.000000  100.000000  100.0   
67    68  100.0  100.000000  100.00000  100.000000  100.000000  100.0   
68    69  100.0   99.328358   99.84127  100.000000  100.000000  100.0   
69    70  100.0  100.000000  100.00000  100.000000  100.000000  100.0   

         Lab 7  Lab 8\n       Lab 9      Lab 10     Lab 11  Lab Total  \
0   100.000000     72.0   74.693878   78.636364  81.617647        480   
1    99.677419     50.0  100.000000   90.000000  90.000000        830   
2   100.000000    100.0  100.000000  100.000000  81.617647        995   
3   100.000000    100.0   50.000000  100.000000  81.617647        930   
4   100.000000    100.0   50.000000   78.636364  81.617647        730   
..         ...      ...         ...         ...        ...        ...   
65  100.000000     72.0   74.693878   80.000000  85.000000        655   
66  100.000000     72.0   74.693878   78.636364  81.617647        500   
67  100.000000     80.0   50.000000   80.000000  81.617647        810   
68  100.000000     72.0   80.000000   10.000000  81.617647        490   
69   99.677419     72.0   74.693878   78.636364  81.617647        500   

    Lab Marks    Record  Lab Internal  Attendance  Total  
0           7  8.000000            13           4     32  
1          12  8.000000            18           4     42  
2          14  8.000000            19           5     46  
3          13  7.000000            16           5     41  
4          10  8.000000            14           4     36  
..        ...       ...           ...         ...    ...  
65          9  7.000000            13           2     31  
66          7  7.000000            16           4     34  
67         12  7.000000            12           4     35  
68          7  7.000000            12           2     28  
69          7  7.835821            17           2     26  

[70 rows x 18 columns]

Now change the data type such that it contains only integer values

In [69]:
df=df.astype('int64')
df

S No  Lab 1  Lab 2  Lab 3  Lab 4  Lab 5  Lab 6  Lab 7  Lab 8\n  Lab 9  \
0      1    100     80    100     99    100    100    100       72     74   
1      2    100    100    100    100     99    100     99       50    100   
2      3    100     95    100    100    100    100    100      100    100   
3      4     80    100    100    100    100    100    100      100     50   
4      5     80    100    100    100     99    100    100      100     50   
..   ...    ...    ...    ...    ...    ...    ...    ...      ...    ...   
65    66    100     99    100    100     90    100    100       72     74   
66    67    100    100     99    100    100    100    100       72     74   
67    68    100    100    100    100    100    100    100       80     50   
68    69    100     99     99    100    100    100    100       72     80   
69    70    100    100    100    100    100    100     99       72     74   

    Lab 10  Lab 11  Lab Total  Lab Marks  Record  Lab Internal  Attendance  \
0       78      81        480          7       8            13           4   
1       90      90        830         12       8            18           4   
2      100      81        995         14       8            19           5   
3      100      81        930         13       7            16           5   
4       78      81        730         10       8            14           4   
..     ...     ...        ...        ...     ...           ...         ...   
65      80      85        655          9       7            13           2   
66      78      81        500          7       7            16           4   
67      80      81        810         12       7            12           4   
68      10      81        490          7       7            12           2   
69      78      81        500          7       7            17           2   

    Total  
0      32  
1      42  
2      46  
3      41  
4      36  
..    ...  
65     31  
66     34  
67     35  
68     28  
69     26  

[70 rows x 18 columns]

<br>
<br>
<br>

## Data Analysis using Pandas

Pandas provides very efficient ways to perform data analysis. Infact, pandas is a tool that is most often used for this puspose.

#### Summary Statistics:
The df.describe method displays multiple summary statistics of the data.

In [70]:
df.describe()

S No       Lab 1       Lab 2       Lab 3       Lab 4       Lab 5  \
count  70.000000   70.000000   70.000000   70.000000   70.000000   70.000000   
mean   35.500000  103.714286   99.314286   99.757143   99.528571   99.171429   
std    20.351085   85.975585    2.936738    1.221020    2.657892    3.530289   
min     1.000000    0.000000   80.000000   90.000000   80.000000   80.000000   
25%    18.250000  100.000000  100.000000  100.000000  100.000000  100.000000   
50%    35.500000  100.000000  100.000000  100.000000  100.000000  100.000000   
75%    52.750000  100.000000  100.000000  100.000000  100.000000  100.000000   
max    70.000000  800.000000  100.000000  100.000000  100.000000  100.000000   

            Lab 6       Lab 7     Lab 8\n       Lab 9      Lab 10     Lab 11  \
count   70.000000   70.000000   70.000000   70.000000   70.000000  70.000000   
mean    98.314286   99.600000   72.000000   74.485714   78.500000  81.300000   
std     11.988573    2.398067   20.387549   16.342124   22.945272   9.473609   
min      0.000000   80.000000    0.000000   50.000000    0.000000  50.000000   
25%    100.000000  100.000000   50.000000   56.000000   78.000000  81.000000   
50%    100.000000  100.000000   72.000000   74.000000   80.000000  81.000000   
75%    100.000000  100.000000   80.000000   90.000000   90.000000  85.000000   
max    100.000000  100.000000  100.000000  100.000000  100.000000  90.000000   

         Lab Total  Lab Marks     Record  Lab Internal  Attendance      Total  
count    70.000000  70.000000  70.000000     70.000000   70.000000  70.000000  
mean    838.785714  11.871429   7.800000     15.571429    4.214286  39.157143  
std     201.863485   2.812608   0.714244      3.132677    1.317583   6.706336  
min     200.000000   3.000000   6.000000      0.000000    0.000000   3.000000  
25%     720.000000  10.000000   7.000000     13.000000    4.000000  36.000000  
50%     902.500000  13.000000   8.000000     16.000000    5.000000  40.500000  
75%     983.750000  14.000000   8.000000     18.000000    5.000000  44.000000  
max    1480.000000  21.000000  10.000000     19.000000    5.000000  48.000000

From the above output, we can conclude that:
- Minimum total marks is: 3
- Maximum total marks is: 48
- on an average the class scored 39.157143

#### Grouping:
The df.groupby() function allows to group data based on a specific column. For example, let us try to find out the average marks students scored based on their attendance.

In [73]:
df.groupby(df['Attendance']).agg({'Total':['mean']})

Total
                 mean
Attendance           
0           21.000000
2           34.125000
4           38.125000
5           41.744186

We see that students with an attendence of 5 scored on an average of 41.744186, those with4 attendence scored 38.125000 and so on. We can see that attendence may be a crucial factor influencing the total marks of the student.

<br>
<br>
<br>

## Applications of Pandas in Data Science:

### Advantages of Numpy:

1. **Easy Data Representation**
- **Pandas**: Uses DataFrames, which are like tables in Excel. They can hold different types of data in each column, making them easy to work with.
- **Traditional Python**: Would require complex combinations of lists or dictionaries, which can be hard to manage.

2. **Simple Data Manipulation**
- **Pandas**: Has built-in functions to easily filter, sort, and group data. You can do complex operations with just a few lines of code.
- **Traditional Python**: need to write more code, often involving loops, which is slower and harder to maintain.

3. **Better Performance**
- **Pandas**: Is optimized to handle large datasets quickly, using the speed of C and Numpy behind the scenes.
- **Traditional Python**: Slower for large datasets, especially when using loops.


5. **Efficient Data Cleaning**
- **Pandas**: Offers tools to handle missing data, remove duplicates, and convert data types easily.
- **Traditional Python**: Requires more manual work, making the process longer and more error-prone.


### Applications in Real World:

- **Finance**: Used for analyzing stock data, managing portfolios, and assessing risks through large datasets of transactions and market prices.

- **Healthcare**: Helps in managing patient records, analyzing clinical trial data, and improving diagnosis and treatment through pattern identification.

- **Social Media**: Performs sentiment and trend analysis by processing large volumes of social media data, helping businesses engage effectively with their audience.

- **Education**: Analyzes student performance data to tailor teaching methods and improve learning outcomes.